pkg/hubble/listener.go

```diff
@@ -0,0 +1,62 @@
+// Copyright 2018-2019 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package hubble
+
+import (
+	"github.com/cilium/hubble/api/v1/flow"
+	"github.com/cilium/hubble/pkg/server"
+	"github.com/gogo/protobuf/types"
+
+	"github.com/cilium/cilium/pkg/monitor/agent/listener"
+	"github.com/cilium/cilium/pkg/monitor/payload"
+	"github.com/cilium/cilium/pkg/node"
+)
+
+type hubbleListener struct {
+	observer server.GRPCServer
+}
+
+// NewHubbleListener returns an initialized pointer to hubbleListener.
+func NewHubbleListener(observer server.GRPCServer) listener.MonitorListener {
+	ml := &hubbleListener{observer}
+	return ml
+}
+
+// Enqueue converts monitor payload to the format accepted by Hubble:
+//   https://github.com/cilium/hubble/blob/04ab72591faca62a305ce0715108876167182e04/api/v1/flow/flow.proto#L266
+func (ml *hubbleListener) Enqueue(pl *payload.Payload) {
+	grpcPl := &flow.Payload{
+		Data:     pl.Data,
+		CPU:      int32(pl.CPU),
+		Lost:     pl.Lost,
+		Type:     flow.EventType(pl.Type),
+		Time:     types.TimestampNow(),
+		HostName: node.GetName(),
+	}
+	select {
+	case ml.observer.GetEventsChannel() <- grpcPl:
+	default:
+		ml.observer.GetLogger().Debug("Per listener queue is full, dropping message")
```

Comment:
So I see this being triggered very often in local testing, even with just a single connection opened (which triggers ~30 events within the same millisecond). This causes Hubble to miss flows.

The issue seems to be that `GetEventsChannel()` buffer size is too small for the "Embedded Mode": By default, the channel returned `GetEventsChannel` will only store 81 flows (`1% of the next power of two of HubbleFlowBufferSize`). This queue size is way to small to deal with short bursts.

The previous buffer size (in Hubble stand-alone mode) has been `GetEventsChannel()` (up to 100 flows) + Cilium monitor's socket queue, which is `1024*NumberOfCPUs` (capped at `16384`).

I will issue a PR on Hubble to fix the queue size of `GetEventsChannel()`. I don't think this PR has to be blocked for this though.

Reply:
thanks @gandro for addressing this in https://github.com/cilium/hubble/pull/163